chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -718,12 +719,36 @@ pub fn validate_chunk_state_witness(
                 child_shard_uid,
             ) => {
                 let old_root = *chunk_extra.state_root();
-                let trie = Trie::from_recorded_storage(
-                    PartialStorage { nodes: transition.base_state },
-                    old_root,
-                    true,
-                );
-                let new_root = trie.retain_split_shard(&boundary_account, retain_mode)?;
+                let partial_storage = PartialStorage { nodes: transition.base_state };
+                let parent_trie = Trie::from_recorded_storage(partial_storage, old_root, true);
+
+                // Update the congestion info based on the parent shard. It's
+                // important to do this step before the `retain_split_shard`
+                // because only the parent has the needed information.
+                if let Some(congestion_info) = chunk_extra.congestion_info_mut() {
+                    // Get the congestion info based on the parent shard.
+                    let epoch_id = epoch_manager.get_epoch_id(&block_hash)?;
+                    let parent_shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+                    let parent_congestion_info = *congestion_info;
+                    *congestion_info = ReshardingManager::get_child_congestion_info_not_finalized(
+                        &parent_trie,
+                        &parent_shard_layout,
+                        parent_congestion_info,
+                        retain_mode,
+                    )?;
+
+                    // Set the allowed shard based on the child shard.
+                    let next_epoch_id = epoch_manager.get_next_epoch_id(&block_hash)?;
+                    let next_shard_layout = epoch_manager.get_shard_layout(&next_epoch_id)?;
+                    ReshardingManager::finalize_allowed_shard(
```

Comment:
This whole bit from `ReshardingManager::get_child_congestion_info_not_finalized` to `ReshardingManager::finalize_allowed_shard` is duplicated across here and resharding manager. I think `parent_trie.retain_split_shard` may be included in the mix. Sounds like a helper function to me!